datafusion-examples/examples/table_sample.rs

```diff
@@ -0,0 +1,1353 @@
+
+        let stmt = statement.clone();
+        match &stmt {
+            Statement::Query(query) => {
+                let inner_plan = sql_to_rel.sql_statement_to_plan(statement)?;
+                if let SetExpr::Select(select) = &*query.body {
+                    if select.from.len() == 1 {
+                        let table_with_joins = select.from.first().unwrap();
+                        if let TableFactor::Table {
+                            sample: Some(table_sample_kind),
+                            ..
+                        } = &table_with_joins.relation
+                        {
+                            debug!("Constructing table sample plan from {:?}", &select);
+                            return self.sample_to_logical_plan(
+                                inner_plan,
+                                table_sample_kind.clone(),
+                            );
+                        }
+                    }
+                }
+                // Pass-through by default
+                Ok(inner_plan)
+            }
+            _ => sql_to_rel.sql_statement_to_plan(statement),
+        }
+    }
+}
+
+// Context provider for tests
+
+struct MockContextProvider<'a> {
+    state: &'a SessionState,
+    tables: HashMap<TableReference, Arc<dyn TableSource>>,
+}
+
+impl ContextProvider for MockContextProvider<'_> {
+    fn get_table_source(&self, name: TableReference) -> Result<Arc<dyn TableSource>> {
+        self.tables
+            .get(&name)
+            .cloned()
+            .ok_or_else(|| plan_datafusion_err!("table '{name}' not found"))
+    }
+
+    fn get_expr_planners(&self) -> &[Arc<dyn ExprPlanner>] {
+        self.state.expr_planners()
+    }
+
+    fn get_type_planner(&self) -> Option<Arc<dyn TypePlanner>> {
+        None
+    }
+
+    fn get_function_meta(&self, name: &str) -> Option<Arc<ScalarUDF>> {
+        self.state.scalar_functions().get(name).cloned()
+    }
+
+    fn get_aggregate_meta(&self, name: &str) -> Option<Arc<AggregateUDF>> {
+        self.state.aggregate_functions().get(name).cloned()
+    }
+
+    fn get_window_meta(&self, name: &str) -> Option<Arc<WindowUDF>> {
+        self.state.window_functions().get(name).cloned()
+    }
+
+    fn get_variable_type(&self, variable_names: &[String]) -> Option<DataType> {
+        self.state
+            .execution_props()
+            .var_providers
+            .as_ref()
+            .and_then(|provider| provider.get(&VarType::System)?.get_type(variable_names))
+    }
+
+    fn options(&self) -> &ConfigOptions {
+        self.state.config_options()
+    }
+
+    fn udf_names(&self) -> Vec<String> {
+        self.state.scalar_functions().keys().cloned().collect()
+    }
+
+    fn udaf_names(&self) -> Vec<String> {
+        self.state.aggregate_functions().keys().cloned().collect()
+    }
+
+    fn udwf_names(&self) -> Vec<String> {
+        self.state.window_functions().keys().cloned().collect()
+    }
+}
+
+#[tokio::main]
```

Comment:
I suggest we move this `main` to the top of the file so people can see what this example does / how it can be used, before having to wade through all the implementation details